### PR TITLE
Add support for race_condition_ttl on fetch

### PR DESCRIPTION
**Pull request by Bobochka**
http://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html

> Setting :race_condition_ttl is very useful in situations where a cache entry is used very frequently and is under heavy load. If a cache expires and due to heavy load several different processes will try to read data natively and then they all will try to write to cache. To avoid that case the first process to find an expired cache entry will bump the cache expiration time by the value set in :race_condition_ttl. Yes, this process is extending the time for a stale value by another few seconds. Because of extended life of the previous cache, other processes will continue to use slightly stale data for a just a bit longer. In the meantime that first process will go ahead and will write into cache the new value. After that all the processes will start getting the new value. The key is to keep :race_condition_ttl small.

### CODE REVIEW

**Comment by tubbo** [comment]
hey @Bobochka sorry for the failing build here, I believe it's related to 'master' still testing on rbx-19mode and jruby-19mode, which has across the board been deprecated by redis-store gems. Can you pull down the latest 'master' and rebase your feature branch against it? That way I can merge this PR. :)

File: lib/active_support/cache/redis_store.rb [hunk 1]
```diff
@@ -58,6 +58,9 @@ def write(name, value, options = nil)
         options = merged_options(options)
         instrument(:write, name, options) do |payload|
           entry = options[:raw].present? ? value : Entry.new(value, options)
+          if options[:expires_in] && options[:race_condition_ttl] && options[:raw].blank?
```
Comment (tubbo) [inline lib/active_support/cache/redis_store.rb]: I would prefer `.present?` checks on each of these options.

File: test/active_support/cache/redis_store_test.rb [hunk 2]
```diff
@@ -351,6 +351,34 @@ def setup
...
+    sleep 1
+
+    th1 = Thread.new do
+      first_rabbit = @store.fetch("rabbit", race_condition_ttl: 2) do
+        sleep 1
+        @white_rabbit
+      end
+    end
+
+    sleep 0.1
```
Comment (tubbo) [inline test/active_support/cache/redis_store_test.rb]: why is this needed?

Comment (Bobochka) [inline test/active_support/cache/redis_store_test.rb]: This is just to assure that first thread already bumped the stale record back to redis, but still hasn't finished generating new value. The other option is to send race_condition_ttl option to second thread as well (or just to have test for both cases).

**Comment by Bobochka** [comment]
@tubbo Thanks for the response. I've added presence checks and rebased onto current master.

**Comment by Bobochka** [comment]
@tubbo I've added one more test. Unfortunately another test without small sleep between threads that we discussed behaves inconsistently. 

Another question that I wanted to ask (not related to this PR) is what do you think about compliance with rails cache interface, since redis-activesupport doesn't always follow it. E.g. 'delete' returns number of deleted keys instead of boolean, 'multi' operations don't check if the entry is expired, etc. Thanks for the response.

**Comment by tubbo** [comment]
@Bobochka We strive to be as compatible as possible with the `ActiveSupport::Cache::Store` interface. That said, since this gem was first introduced there have been quite a few changes to that interface. It's possible the interface for `Cache::Store#delete` changed over the course of the years, so we would be happy to have any contributions that keep the redis-store gems consistent with the interface.
